src/language/branches.md

```diff
@@ -0,0 +1,295 @@
+    // Matches if shape is of type Rect, and then against the properties of Rect
+  }
+  ```  
+
+Values destructured by a case pattern become local variables, only available in the
+scope of that clause. 
+
+### Switch statements
+
+Non-empty `case` clauses implicitly jump to the end of the switch after completion.
+Other valid ways to end a non-empty `case` clause are a [`continue`][break],
+[`throw`][], or [`return`][] statement.
+
+Use a `default` or wildcard `_` clause to execute code when no `case` clause matches:
+
+```dart
+var command = 'OPEN';
+switch (command) {
+  case 'CLOSED':
+    executeClosed();
+  case 'PENDING':
+    executePending();
+  case 'APPROVED':
+    executeApproved();
+  case 'DENIED':
+    executeDenied();
+  case 'OPEN':
+    executeOpen();
+  default:
+    executeUnknown();
+}
+```
+
+Empty cases fall through to the next case. For an empty case that does not fall
+through, use [`break`][break] for its body:
+
+```dart
+var command = 'CLOSED';
+switch (command) {
+  case 'CLOSED': // Empty case falls through.
+  case 'NOW_CLOSED':
+    // Runs for both CLOSED and NOW_CLOSED.
+    executeNowClosed();
+}
+```
+
+For non-sequential fall-through, you can use a [`continue` statement][break] and
+a label:
+
+```dart
+var command = 'CLOSED';
+switch (command) {
+  case 'CLOSED':
+    executeClosed();
+    continue nowClosed;
+  // Continues executing at the nowClosed label.
+
+  nowClosed:
+  case 'NOW_CLOSED':
+    // Runs for both CLOSED and NOW_CLOSED.
+    executeNowClosed();
+}
+```
+
+### Switch expressions
+
+_Switch expressions_ write a result value determined by a switch
+directly into a variable or return statement. 
+
+The syntax rules for a switch expression are:
+- The expression body is a series of case patterns.
+- Cases _do not_ use the `case` keyword.
+- Case patterns are followed by `=>`, and then a case body. 
+- Case bodies terminate with a comma `,`.
+- The switch expression body terminates with a semicolon `;`.
+- The wildcard pattern `_` catches remaining values instead of `default`.
+
+This allows you to rewrite a 
+_statement_ like this:
+
+```dart
+switch (charCode) {
+  case slash when nextCharCode == slash:   // Constant pattern with guard
+    skipComment();
+  case slash || star || plus || minus:     // Logical-or pattern
+    operator(charCode);
+  case >= digit0 && <= digit9:             // Relational and logical-and patterns
+    number();
+  default:
+    invalid();
+}
+```
+
+Into an _expression_, like this:
+
+```dart
+var token = switch (charCode) {
+  slash when nextCharCode == slash => skipComment(),
+  slash || star || plus || minus   => operator(charCode),
+  >= digit0 && <= digit9           => number(),
+  _                                => throw invalid()
+};
+```
+
+### Exhaustiveness checking
+
+Exhaustiveness checking is a feature of switches that raises a compile-time error
+if a switch is not exhaustively covering all possible values that could flow through it.
```

Comment:
Defining exhaustiveness in terms of "exhaustively" may not be very clear. :)

Maybe "reports a compile-time error if it's possible for a value entering the switch to not match any of the case patterns".

Reply:
I went with your version, except removed "patterns" from the end (just "cases") because of [Erik's idea](https://github.com/dart-lang/site-www/pull/4780#discussion_r1176761413) to not get into "case patterns" terminology too much here